Mathlib/Topology/Connected/RealIntervals.lean

```diff
@@ -0,0 +1,560 @@
+theorem classify_connected_bounded_reals_nonempty_interior
+    {X : Set ℝ} (conn : IsConnected X) (above : BddAbove X) (below : BddBelow X)
+    (h : (interior X) ≠ ∅) :
+    isBoundedInterval X := by
+  rcases classify_connected_bounded_reals conn above below with ⟨a, ha⟩ | bdd
+  · rw [ha, interior_singleton] at h
+    exact (h rfl).elim
+  · exact bdd
+
+/-! ### Facts about `BoundedInterval`s -/
+
+section I_section
+variable
+  (I : BoundedInterval)
+
+lemma isConnected_BoundedInterval :
+    IsConnected (BoundedInterval_as_set I) := by
+  rw [BoundedInterval_as_set]
+  have ⟨a, b, lt, kind⟩ := I
+  match kind with
+  | IooKind => exact isConnected_Ioo lt
+  | IocKind => exact isConnected_Ioc lt
+  | IcoKind => exact isConnected_Ico lt
+  | IccKind => exact isConnected_Icc (LT.lt.le lt)
+
+lemma boundedInterval_nonempty : (BoundedInterval_as_set I).Nonempty :=
+  IsConnected.nonempty (isConnected_BoundedInterval I)
+
+lemma BoundedInterval_subset_Icc :
+    BoundedInterval_as_set I ⊆ Icc I.left_endpoint I.right_endpoint := by
+  have ⟨a, b, lt, kind⟩ := I
+  match kind with
+  | IooKind => dsimp; exact Ioo_subset_Icc_self
+  | IocKind => dsimp; exact Ioc_subset_Icc_self
+  | IcoKind => dsimp; exact Ico_subset_Icc_self
+  | IccKind => dsimp; exact Eq.subset rfl
+
+lemma BoundedInterval_contains_Ioo :
+    Ioo I.left_endpoint I.right_endpoint ⊆ BoundedInterval_as_set I := by
+  have ⟨a, b, lt, kind⟩ := I
+  match kind with
+  | IooKind => dsimp; exact Eq.subset rfl
+  | IocKind => dsimp; exact Ioo_subset_Ioc_self
+  | IcoKind => dsimp; exact Ioo_subset_Ico_self
+  | IccKind => dsimp; exact Ioo_subset_Icc_self
+
+lemma closure_BoundedInterval :
+    closure (BoundedInterval_as_set I) = Icc I.left_endpoint I.right_endpoint := by
+  apply Subset.antisymm
+  · calc closure (BoundedInterval_as_set I)
+      ⊆ closure (Icc I.left_endpoint I.right_endpoint) :=
+        closure_mono (BoundedInterval_subset_Icc I)
+    _ = Icc I.left_endpoint I.right_endpoint :=
+      IsClosed.closure_eq isClosed_Icc
+  · calc Icc I.left_endpoint I.right_endpoint
+      = closure (Ioo I.left_endpoint I.right_endpoint) :=
+        (closure_Ioo (ne_of_lt I.left_lt_right)).symm
+    _ ⊆ closure (BoundedInterval_as_set I) :=
+      closure_mono (BoundedInterval_contains_Ioo I)
+
+lemma isBoundedBelow_BoundedInterval : BddBelow (BoundedInterval_as_set I) :=
+  BddBelow.mono (BoundedInterval_subset_Icc I) bddBelow_Icc
+
+lemma isBoundedAbove_BoundedInterval : BddAbove (BoundedInterval_as_set I) :=
+  BddAbove.mono (BoundedInterval_subset_Icc I) bddAbove_Icc
+
+theorem boundedInterval_left_endpoint_isGLB :
+    IsGLB (BoundedInterval_as_set I) I.left_endpoint := by
+  let icc := Icc I.left_endpoint I.right_endpoint
+  let ioo := Ioo I.left_endpoint I.right_endpoint
+  have h₁ : IsGLB icc I.left_endpoint := by
+    rw [← csInf_Icc (le_of_lt I.left_lt_right)]
+    apply Real.is_glb_sInf icc
+    exact nonempty_Icc.mpr (le_of_lt I.left_lt_right)
```

Comment:
Two style comments:
- your apply creates two subgoals; we try to use focusing dots in this case
(Rule of thumb: there should be only one top-level goal.)
- you could solve the second goal by doing something like
    `refine Real.is_glb_sInf icc ?_ bddBelow_Icc` or perhaps `apply Real.is_glb_sInf icc _ bddBelow_Icc` (untested). Then you only have one goal left.
-